File: te1.r
library(copent) # transfer entropy (TE)
library(CondIndTests) # kernel-based CI test (KCI)
library(RCIT) # Randomized conditional Correlation Test (RCoT)
library(cdcsis) # conditional distance correlation (CDC)
library(FOCI) # conditional dependence coefficient (CODEC)
library(GeneralisedCovarianceMeasure) # Generalised Covariance Measure (GCM)
library(weightedGCM) # weighted GCM
library(KPC) # Kernel Partial Correlation (KPC)
library(ppcor) # Partial Correlation (pcor)
library(CondCopulas) # Conditional Kendall's Tau (CKT)
library(EDMeasure) # Conditional Mean Dependence (CMD)
source("https://raw.githubusercontent.com/lassepetersen/partial-copula-CI-test/main/parCopCITest.R") # partial copula based CI test

prsa2010data = read.csv("https://archive.ics.uci.edu/ml/machine-learning-databases/00381/PRSA_data_2010.1.1-2014.12.31.csv")
# id: 6(PM2.5), 7(Dew Point), 8(Temperature), 9(Pressure), 11(Cumulative Wind Speed)
idx = c(6,9)
data = prsa2010data[2200:2700, idx]

te1 = 0
kci1 = 0
rcot1 = 0
cdc1 = 0
codec1 = 0
gcm1 = 0
wgcm1 = 0
kpc1 = 0
pcor1 = 0
ckt1 = 0
cmd1 = 0
pcop1 = 0
for (lag in 1:24){
  pm25a = data[1:(501-lag),1]
  pm25b = data[(lag+1):501,1]
  v1 = data[1:(501-lag),2]
  
  # conditional independence test with copula entropy
  te1[lag] = ci(pm25b,v1,pm25a)  
  # or estimating transfer entropy directly as below
  # te1[lag] = transent(data[,1],data[,2],lag)
  
  kci1[lag] = KCI(pm25b,v1,pm25a)$testStatistic
  rcot1[lag] = RCIT(pm25b,v1,pm25a)$Sta
  cdc1[lag] = cdcor(pm25b,v1,pm25a)$statistic
  codec1[lag] = codec(pm25b,v1,pm25a)
  gcm1[lag] = gcm.test(pm25b,v1,pm25a)$test.statistic
  wgcm1[lag] = wgcm.est(pm25b,v1,pm25a, regr.meth = "xgboost")
  kpc1[lag] = KPCRKHS(pm25b,pm25a,v1)
  pcor1[lag] = pcor.test(pm25b,v1,pm25a)$statistic
  ckt1[lag] = CKT.estimate(pm25b,v1,pm25a, methodEstimation = "tree", h = 0.1)
  cmd1[lag] = cmdm_test(pm25b,v1,pm25a)$stat
  pcop1[lag] = test_CI(pm25b,v1,pm25a)$statistic
}

# TE via CE
x11()
plot(te1, xlab = "lag (hours)", ylab = "Transfer Entropy", main = "Pressure")
lines(te1)
# KCI
x11()
plot(kci1, xlab = "lag (hours)", ylab = "KCI", main = "Pressure")
lines(kci1)
# RCoT
x11()
plot(rcot1, xlab = "lag (hours)", ylab = "RCoT", main = "Pressure")
lines(rcot1)
# CDC
x11()
cdc1 = unlist(cdc1)
plot(cdc1, xlab = "lag (hours)", ylab = "CDC", main = "Pressure")
lines(cdc1)
# CODEC
x11()
plot(codec1, xlab = "lag (hours)", ylab = "CODEC", main = "Pressure")
lines(codec1)
# GCM
x11()
plot(gcm1, xlab = "lag (hours)", ylab = "GCM", main = "Pressure")
lines(gcm1)
# wGCM
x11()
plot(wgcm1, xlab = "lag (hours)", ylab = "wGCM", main = "Pressure")
lines(wgcm1)
# KPC
x11()
plot(kpc1, xlab = "lag (hours)", ylab = "KPC", main = "Pressure")
lines(kpc1)
# Partial Correlation
x11()
pcor1 = unlist(pcor1)
plot(pcor1, xlab = "lag (hours)", ylab = "Partial Correlation", main = "Pressure")
lines(pcor1)
# Conditional Kendall's Tau
x11()
plot(ckt1, xlab = "lag (hours)", ylab = "CKT", main = "Pressure")
lines(ckt1)
# Conditional Mean Dependence
x11()
plot(cmd1, xlab = "lag (hours)", ylab = "CMD", main = "Pressure")
lines(cmd1)
# partial copula based CI test
x11()
plot(pcop1, xlab = "lag (hours)", ylab = "PartialCopula", main = "Pressure")
lines(pcop1)

File: te1.py
from copent import transent, ci
from knncmi import cmi
from causallearn.utils.KCI.KCI import KCI_CInd
from pycit import citest
from fcit import fcit
from CCIT import CCIT
from pcit.IndependenceTest import PCIT  ##  the 'mlxtend' package required
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00381/PRSA_data_2010.1.1-2014.12.31.csv"
prsa2010 = pd.read_csv(url)
# index: 5(PM2.5),6(Dew Point),7(Temperature),8(Pressure),10(Cumulative Wind Speed)
varname = ["","","","","","PM2.5","Dew Point","Temperature","Pressure","","CWS"]
eid = 5 # effect
cid = 8 # cause
data = prsa2010.iloc[2200:2700,[eid,cid]].values

te1 = np.zeros(24)
cmi1 = np.zeros(24)
kci1 = np.zeros(24)
ci1 = np.zeros(24)
fcit1 = np.zeros(24)
ccit1 = np.zeros(24)
pcit1 = np.zeros(24)
for lag in range(1,25):
	x1 = data[0:(500-lag),0]
	x2 = data[lag:500,0]
	y = data[0:(500-lag),1]
	
	## copent
	te1[lag-1] = ci(x2,y,x1)
	# te1[lag-1] = transent(data[:,0],data[:,1],lag)
	## knncmi
	df = pd.DataFrame(np.vstack((x2,y,x1)).T, columns = ['x2','y','x1'])
	cmi1[lag-1] = cmi(['x2'], ['y'], ['x1'], 3, df)	
	## pycit
	ci1[lag-1] = citest(x2, y, x1, test_args={'statistic': 'ksg_mi'})
	## fcit
	len1 = len(x1)
	x2a = np.reshape(x2,[len1,1])
	ya = np.reshape(y,[len1,1])
	x1a = np.reshape(x1,[len1,1])
	fcit1[lag-1] = fcit.test(x2a,ya,x1a)
	## causal-learn
	kci_cind = KCI_CInd()
	kci1[lag-1],_ = kci_cind.compute_pvalue(x2a,ya,x1a)
	## CCIT
	ccit1[lag-1] = CCIT.CCIT(x2a,ya,x1a)
	## pcit
	p,tmp,tmp = PCIT(x2a,ya,z = x1a)
	pcit1[lag-1] = p[0]
	
	str = "lag: %d; TE:%.3f; CMI:%.3f; KCI:%.3f; CI:%.3f; fcit:%.3f; ccit:%.3f; pcit:%.3f" %(lag,te1[lag-1],cmi1[lag-1],kci1[lag-1],ci1[lag-1],fcit1[lag-1],ccit1[lag-1],pcit1[lag-1])
	print(str)
	
plt.plot(te1,'g',label = 'copent')
plt.plot(cmi1,'r',label = 'knncmi')
plt.plot(kci1,'c', label = 'causal-learn')
plt.plot(ci1,'b', label = 'pycit')
plt.plot(fcit1,'y', label = 'fcit')
plt.plot(ccit1,'m', label = 'ccit')
plt.plot(pcit1,'k', label = 'pcit')
title = "%s on %s" % (varname[cid],varname[eid])
plt.title(title)
plt.xlabel('lag(hours)')
pos = np.array([4,9,14,19])
plt.xticks(pos,pos+1)
plt.ylabel('Causality')
plt.legend()
plt.show()		
